File: go/backoff.go
// SPDX-License-Identifier: MIT
package geyserlite

import "time"

// backoff is a non-thread-safe exponential backoff used by the subprocess
// supervisor between restart attempts. Doubles each next() call, caps at max,
// resets on a clean run.
type backoff struct {
	min, max, cur time.Duration
}

func newBackoff(minD, maxD time.Duration) *backoff {
	if minD <= 0 {
		minD = time.Second
	}
	if maxD <= 0 || maxD < minD {
		maxD = 60 * time.Second
	}
	return &backoff{min: minD, max: maxD, cur: minD}
}

func (b *backoff) next() time.Duration {
	d := b.cur
	b.cur *= 2
	if b.cur > b.max {
		b.cur = b.max
	}
	return d
}

func (b *backoff) peek() time.Duration { return b.cur }

func (b *backoff) reset() { b.cur = b.min }

File: rust/src/backoff.rs
// SPDX-License-Identifier: MIT
use std::time::Duration;

/// Exponential backoff used by the subprocess supervisor between restart attempts.
pub(crate) struct Backoff {
    min: Duration,
    max: Duration,
    cur: Duration,
}

impl Backoff {
    pub fn new(min: Duration, max: Duration) -> Self {
        let min = if min.is_zero() {
            Duration::from_secs(1)
        } else {
            min
        };
        let max = if max < min {
            Duration::from_secs(60)
        } else {
            max
        };
        Self { min, max, cur: min }
    }
    pub fn next(&mut self) -> Duration {
        let d = self.cur;
        self.cur = (self.cur * 2).min(self.max);
        d
    }
    pub fn reset(&mut self) {
        self.cur = self.min;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn exponential_with_cap() {
        let mut b = Backoff::new(Duration::from_secs(1), Duration::from_secs(8));
        assert_eq!(b.next(), Duration::from_secs(1));
        assert_eq!(b.next(), Duration::from_secs(2));
        assert_eq!(b.next(), Duration::from_secs(4));
        assert_eq!(b.next(), Duration::from_secs(8)); // capped
        assert_eq!(b.next(), Duration::from_secs(8));
    }

    #[test]
    fn reset_returns_to_min() {
        let mut b = Backoff::new(Duration::from_secs(1), Duration::from_secs(8));
        b.next();
        b.next();
        b.next();
        b.reset();
        assert_eq!(b.next(), Duration::from_secs(1));
    }

    #[test]
    fn zero_inputs_get_sane_defaults() {
        let mut b = Backoff::new(Duration::ZERO, Duration::ZERO);
        assert!(b.next() > Duration::ZERO);
    }
}
